File: ios/Plugin/Adapters/PrinterAdapter.swift
import Foundation
import UIKit

/// Contrat commun à tous les adapters iOS (miroir de l'interface Android).
///
/// Toutes les méthodes longues sont `async` et lèvent `PrinterError` en cas d'échec.
protocol PrinterAdapter {
    var id: AdapterId { get }

    /// True si le SDK requis est lié à l'app (sinon adapter ignoré).
    func isAvailable() -> Bool

    /// Découverte propre à l'adapter, résultats poussés via `onFound`.
    func discover(timeoutMs: Int, onFound: @escaping (DiscoveredPrinter) -> Void) async

    func canHandle(_ profile: PrinterProfile) -> Bool

    func connect(_ profile: PrinterProfile, timeoutMs: Int) async throws
    func isConnected(_ printerId: String) -> Bool
    func disconnect(_ printerId: String) async

    /// Imprime un UIImage DÉJÀ redimensionné à la largeur cible.
    /// Retourne le nombre d'octets envoyés (best effort).
    func printImage(_ profile: PrinterProfile, image: UIImage, options: RenderOptions) async throws -> Int

    func getStatus(_ profile: PrinterProfile) async throws -> PrinterStatus
}

File: android/src/main/java/com/resto/thermalprinter/adapters/PrinterAdapter.kt
package com.resto.thermalprinter.adapters

import android.graphics.Bitmap
import com.resto.thermalprinter.model.AdapterId
import com.resto.thermalprinter.model.DiscoveredPrinter
import com.resto.thermalprinter.model.PrinterProfile
import com.resto.thermalprinter.model.PrinterStatus
import com.resto.thermalprinter.model.RenderOptions

/**
 * Contrat commun à tous les adapters natifs Android.
 *
 * Un adapter encapsule TOUT ce qui est spécifique à une famille d'imprimantes :
 *   - sa découverte (via SDK ou scan générique),
 *   - sa connexion / reconnexion,
 *   - la conversion d'un Bitmap déjà binarisé en commandes natives,
 *   - l'envoi,
 *   - la lecture de statut.
 *
 * Le moteur (ThermalPrinterEngine) orchestre les adapters et applique la priorité.
 * Toutes les méthodes longues sont `suspend` (exécutées sur Dispatchers.IO).
 */
interface PrinterAdapter {

    val id: AdapterId

    /** True si le SDK requis est présent dans l'app (sinon l'adapter est ignoré). */
    fun isAvailable(): Boolean

    /**
     * Lance une découverte propre à cet adapter. Les résultats partiels peuvent
     * être émis via [onFound] au fil de l'eau. La méthode se termine quand le
     * scan est fini ou que [timeoutMs] est atteint.
     */
    suspend fun discover(
        timeoutMs: Long,
        onFound: (DiscoveredPrinter) -> Unit,
    )

    /** Indique si cet adapter sait gérer ce profil (transport + identité). */
    fun canHandle(profile: PrinterProfile): Boolean

    /** Ouvre une connexion. Idempotent : ne rien faire si déjà connecté. */
    suspend fun connect(profile: PrinterProfile, timeoutMs: Long)

    /** True si une connexion active existe pour cette imprimante. */
    fun isConnected(printerId: String): Boolean

    /** Ferme la connexion. */
    suspend fun disconnect(printerId: String)

    /**
     * Imprime un bitmap.
     *
     * IMPORTANT : `bitmap` est DÉJÀ redimensionné à la largeur cible et
     * binarisé/dithered par le moteur (ImageProcessor). Pour les adapters ESC/POS,
     * il suffit de l'encoder en raster GS v 0. Pour les SDK fabricants, on passe
     * le bitmap à l'API d'impression image du SDK.
     *
     * @return nombre d'octets envoyés (best effort).
     */
    suspend fun printBitmap(profile: PrinterProfile, bitmap: Bitmap, options: RenderOptions): Int

    /** Lit le statut. Renvoie supportsStatus=false via online/paper=unknown si non supporté. */
    suspend fun getStatus(profile: PrinterProfile): PrinterStatus
}
